### ISSUE TITLE
Differences in label rendering between local and production for OSM-carto

### ISSUE THREAD

**Issue by yohanboniface** [kosmtik/kosmtik#57]
Refs:
- https://github.com/gravitystorm/openstreetmap-carto/issues/657#issuecomment-78665177
- https://lists.openstreetmap.org/pipermail/dev/2015-April/028504.html

**Comment by pnorman** [comment]
The biggest and most obvious difference is going to be metatile size.

For reference, tile.osm.org uses

Mapnik 2.2
renderd+mod_tile
8x8 metatiles

**Comment by yohanboniface** [comment]
Metatile size seems a good point!

@math1985 @nebulon42 are you using the [metatile size from project.yaml](https://github.com/gravitystorm/openstreetmap-carto/blob/master/project.yaml#L2) or are you changing the value on your localconfig?

**Comment by nebulon42** [comment]
First I used the value from project.yaml and then I tried to change it in the local config with `project.mml.metatile`. The problems remain. I will now try to use different versions of Mapnik.

How does Kosmtik handle metatile buffers? (if that could have anything to do with it)

**Comment by yohanboniface** [comment]
> How does Kosmtik handle metatile buffers? (if that could have anything to do with it)

Kosmtik doesn't deal with this directly, as, AFAIK, it's only set through the carto directive `buffer-size`. So this is directly handled by carto itself.

**Comment by nebulon42** [comment]
> set through the carto directive `buffer-size`

I have now tried setting `buffer-size: 256` for the Map object, but this does not help. I still have to try other versions of Mapnik.

**Comment by nebulon42** [comment]
I upgraded Mapnik by executing `npm install mapnik@3.1.2` which did not change the situation.

**Comment by math1985** [comment]
What was the old version? The default Kosmtik version is likely newer than the default Tilemill version.

**Comment by nebulon42** [comment]
The default version should have been 2.3.0 according to the output on the command line. Unfortunately, I cannot compare with Tilemill because I got compatibility issues with my node.js package.

**Comment by pnorman** [comment]
> I upgraded Mapnik by executing `npm install mapnik@3.1.2` which did not change the situation

this should upgrade *node-mapnik* but I don't think it upgrades mapnik.

**Comment by nebulon42** [comment]
Of course you are right. But since I do not have any other mapnik packages installed on my system it somehow includes also mapnik binaries. Or maybe I'm missing something more fundamental..as I said my knowledge about mapnik is very close to zero.

**Comment by nebulon42** [comment]
@yohanboniface I'm seeing the same artefacts with Mapbox' osm-bright, although not so many. [Their mml](https://github.com/mapbox/osm-bright/blob/master/osm-bright/osm-bright.osm2pgsql.mml) has no metatile size set. I'm using the default versions of libraries.

**Comment by yohanboniface** [comment]
@tomhughes @pnorman Can you maybe provide us the Mapnik XML from production server (cleaned from its passwords if any of course), so we can compare with the local ones and have a better idea of the differences, and then start investigating from there? Thanks :)

**Comment by tomhughes** [comment]
Well the production servers are using mapnik 2.2 so if you're using 2.3 or 3.x then that might explain the differences?

**Comment by yohanboniface** [comment]
> Well the production servers are using mapnik 2.2 so if you're using 2.3 or 3.x then that might explain
> the differences?

Yes, that's certainly the explanation, but comparing the two computed XML could help confirm that, or disconfirm. :)

**Comment by tomhughes** [comment]
Well that won't alter the XML will it - the XML is input to mapnik. Only the carto version will alter that.

Currently both machines are running carto@0.14.1 which I believe is the current release version.

**Comment by nebulon42** [comment]
It seems to me that the problem with cutoff labels in openstreetmap-carto is now gone. Can anybody confirm this?

**Comment by math1985** [comment]
I still experience this problem. I didn't update my Kosmtik, so it might have been fixed in Kosmtik, but at least it hasn't been fixed on the style side.

**Comment by nebulon42** [comment]
@math1985 I'm running the HEAD version of Kosmtik. Maybe you could do an update (in a fresh clone?) to see if the problem was fixed by something on the side of Kosmtik or with the new Mapnik node packages.

**Comment by nebulon42** [comment]
Closer inspection revealed that the problem is still there, but less visible.

**Comment by pnorman** [comment]
Using HEAD Kosmtik will *increase* the differences as it will pull in a node-mapnik that depends on Mapnik 3. It will probably reduce the number of cut-off labels, but the problem isn't cut-off label, it's differences.

If the Metatile size is set to 8x8 for the style, are there still differences?
